### PR TITLE
5 new financial functions

### PR DESCRIPTION
**Pull request by izulin**
### Context

Introduces 5 new financial functions: FVSCHEDULE, NPV, MIRR, PDURATION, XNPV.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature or improvement (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Additional language file or change to the existing one (translations)

### Related issue(s):
1.
2.
3.

### Checklist:

- [x] My code follows the code style of this project,
- [ ] My change requires a change to the documentation,
- [x] I described the modification in the CHANGELOG.md file.

### CODE REVIEW

File: src/interpreter/plugin/FinancialPlugin.ts [hunk 1]
```diff
@@ -559,6 +599,112 @@ export class FinancialPlugin extends FunctionPlugin {
...
+    return this.runFunction(ast.args, formulaAddress, this.metadata('PDURATION'),
+      (rate: number, pv: number, fv: number) => (Math.log(fv) - Math.log(pv))/Math.log(1+rate)
+    )
+  }
+
+  public xnpv(ast: ProcedureAst, formulaAddress: SimpleCellAddress): InternalScalarValue {
+    return this.runFunction(ast.args, formulaAddress, this.metadata('XNPV'),
+      (rate: number, values: SimpleRangeValue, dates: SimpleRangeValue) => {
+        const valArr = values.valuesFromTopLeftCorner()
+        for(let val of valArr) {
```
Comment (izulin) [inline src/interpreter/plugin/FinancialPlugin.ts:676]:
```suggestion
        for(const val of valArr) {
```

File: src/interpreter/plugin/FinancialPlugin.ts [hunk 2]
```diff
@@ -559,6 +599,112 @@ export class FinancialPlugin extends FunctionPlugin {
...
+          if(typeof val !== 'number') {
+            return new CellError(ErrorType.NUM, ErrorMessage.NumberExpected)
+          }
+        }
+        const valArrNum = valArr as number[]
+        const dateArr = dates.valuesFromTopLeftCorner()
+        for(let date of dateArr) {
```
Comment (izulin) [inline src/interpreter/plugin/FinancialPlugin.ts:683]:
```suggestion
        for(const date of dateArr) {
```

File: test/interpreter/function-npv.spec.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+    ])
+    expect(engine.getCellValue(adr('A1'))).toEqual(detailedError(ErrorType.NA))
+    expect(engine.getCellValue(adr('A2'))).toEqual(0.5)
+    expect(engine.getCellValue(adr('A3'))).toEqual(0.5)
+    expect(engine.getCellValue(adr('A4'))).toEqual(detailedError(ErrorType.VALUE, ErrorMessage.NumberCoercion))
+    expect(engine.getCellValue(adr('A5'))).toEqual(0.75)
+    expect(engine.getCellValue(adr('A6'))).toEqual(0)
+    expect(engine.getCellValue(adr('A7'))).toEqual(detailedError(ErrorType.NUM, ErrorMessage.NaN))
+  })
+
```
Comment (wojciechczerniak) [inline test/interpreter/function-npv.spec.ts:32]: - Can we add a test with a range 3x3? There is this order condition we should check "Because the result is affected by the order of values, evaluators shall evaluate arguments in the order given and range reference and array arguments row-wise starting from top left." [1]

I think other conditions are here but not that clearly described

Source: 
[1] https://docs.oasis-open.org/office/OpenDocument/v1.3/csprd02/part4-formula/OpenDocument-v1.3-csprd02-part4-formula.html#__RefHeading__1018290_715980110
[2] https://support.microsoft.com/en-us/office/npv-function-8672cb67-2576-4d07-b67b-ac28acf2a568

File: test/interpreter/function-mirr.spec.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+      ['=MIRR(B3:E3,0.2,0.1)', -1, 0, -1, 1],
+      ['=MIRR(B4:E4,0.2,0.1)', -1, 0, -1, -1],
+      ['=MIRR(B5:G5,0.2,0.1)', -1, 0, 'abcd', true, -1, 1],
+    ])
+    expect(engine.getCellValue(adr('A1'))).toEqual(3)
+    expect(engine.getCellValue(adr('A2'))).toEqual(0)
+    expect(engine.getCellValue(adr('A3'))).toBeCloseTo(-0.161201673643132, 6)
+    expect(engine.getCellValue(adr('A4'))).toEqual(detailedError(ErrorType.DIV_BY_ZERO))
+    expect(engine.getCellValue(adr('A5'))).toBeCloseTo(-0.161201673643132, 6)
+  })
```
Comment (wojciechczerniak) [inline test/interpreter/function-mirr.spec.ts:29]: All the rules [1,2] are tested but please describe the tests and conditions

```suggestion
  it('should return correct value', () => {
    const engine = HyperFormula.buildFromArray([
      ['=MIRR(B1:C1,1,1)', 1, -1],
      ['=MIRR(B2:C2,1,1)', -1, 1],
      ['=MIRR(B3:E3,0.2,0.1)', -1, 0, -1, 1],
      ['=MIRR(B3:E3,0.2,0.1)', -1, -1, 1],
    ])
    expect(engine.getCellValue(adr('A1'))).toEqual(3)
    expect(engine.getCellValue(adr('A2'))).toEqual(0)
    expect(engine.getCellValue(adr('A3'))).toBeCloseTo(-0.161201673643132, 6)
    expect(engine.getCellValue(adr('A4'))).toBeCloseTo(-0.161201673643132, 6) // different value without 0
  })
  
  it('should return #DIV/0! if at least one positive and one negative values condtion is not met', () => {
    const engine = HyperFormula.buildFromArray([
      ['=MIRR(B1:E1,0.2,0.1)', -1, 0, -1, -1],
      ['=MIRR(B2:E2,0.2,0.1)', 1, 0, 1, 1],
    ])
    expect(engine.getCellValue(adr('A1'))).toEqual(detailedError(ErrorType.DIV_BY_ZERO))
    expect(engine.getCellValue(adr('A2'))).toEqual(detailedError(ErrorType.DIV_BY_ZERO))
  })
  
  it('should ignore text, boolean and empty values', () => {
    const engine = HyperFormula.buildFromArray([
      ['=MIRR(B1:H1,0.2,0.1)', -1, 0, 'abcd', true, null, -1, 1],
    ])

    expect(engine.getCellValue(adr('A1'))).toBeCloseTo(-0.161201673643132, 6)
  })
```

Source:
[1] https://support.microsoft.com/en-us/office/mirr-function-b020f038-7492-4fb4-93c1-35c345b53524
[2] https://docs.oasis-open.org/office/OpenDocument/v1.3/csprd02/part4-formula/OpenDocument-v1.3-csprd02-part4-formula.html#__RefHeading__1018284_715980110

File: test/interpreter/function-fvschedule.spec.ts [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+      ['=FVSCHEDULE(1, B5)', 'abcd']
+    ])
+
+    expect(engine.getCellValue(adr('A1'))).toEqual(2)
+    expect(engine.getCellValue(adr('A2'))).toEqual(8)
+    expect(engine.getCellValue(adr('A3'))).toEqual(8)
+    expect(engine.getCellValue(adr('A4'))).toEqual(detailedError(ErrorType.NA))
+    expect(engine.getCellValue(adr('A5'))).toEqual(detailedError(ErrorType.VALUE, ErrorMessage.NumberCoercion))
+  })
+})
```
Comment (wojciechczerniak) [inline test/interpreter/function-fvschedule.spec.ts:30]: Can `Principal` be negative?
Empty cells are a valid value, test missing.
Boolean is a valid value? What does Excel do in this case? The description says: "The values in schedule can be numbers or blank cells; any other value produces the #VALUE! error"

https://docs.oasis-open.org/office/OpenDocument/v1.3/csprd02/part4-formula/OpenDocument-v1.3-csprd02-part4-formula.html#__RefHeading__1018272_715980110

**Review by wojciechczerniak — CHANGES_REQUESTED** [review]

File: test/interpreter/function-xnpv.spec.ts [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+    ])
+
+    expect(engine.getCellValue(adr('A1'))).toEqual(1)
+    expect(engine.getCellValue(adr('A2'))).toBeCloseTo(2.99620553730319, 6)
+    expect(engine.getCellValue(adr('A3'))).toBeCloseTo(2.99620553730319, 6)
+    expect(engine.getCellValue(adr('A4'))).toEqual(detailedError(ErrorType.NUM, ErrorMessage.ValueSmall))
+    expect(engine.getCellValue(adr('A5'))).toEqual(detailedError(ErrorType.NUM, ErrorMessage.NumberExpected))
+    expect(engine.getCellValue(adr('A6'))).toEqual(detailedError(ErrorType.NUM, ErrorMessage.ValueSmall))
+  })
+})
```
[the remark this answers is not in the document]
Comment (izulin) [reply test/interpreter/function-xnpv.spec.ts:32]: In general I implemented what is implemented in Excel, not what the documentation states. See e.g. discrepancy between logic in NPV (documentation vs implementation).

> ODFF [1] says "Rate: discount rate. The value should be greater than -1. "

Excel accepts XNPV rate only if its greater than 0. 

> Missing test for "If any number in dates precedes the starting date, XNPV returns the #NUM! error value." [2]

It is tested.

> Missing test where Rate is non-numeric: "If any argument is nonnumeric, XNPV returns the #VALUE! error value."

> I think that "If any number in dates is not a valid date, XNPV returns the #VALUE! error value." depends on date configuration. We should add tests, where different nullDate is set and the date is not parsable (would be negative value) or is above the max date setting.

It makes no sense to test functionality that is part of the wrapper, not of the function actually implemented. E.g. parsing of dates into numbers is fully out of control of the function, and testing it here does not test the function (which actually takes as an input array of scalars).

> Both Excel "The series of values must contain at least one positive value and one negative value." and ODFF "The series of cash flow should contain at least one positive and one negative value. " agree that there should be a positive and negative value. But i get a result anyway. We have to investigate this

It's not implemented in Excel like that, hence my implementation follows what actually is implemented, not what is documented.

File: test/interpreter/function-npv.spec.ts [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+  /**
+   * Discrepancy with products #1 and #2.
+   */
+  it('cell reference', () => {
+    const engine = HyperFormula.buildFromArray([
+      ['=NPV(1,B1)', true]
+    ])
+    expect(engine.getCellValue(adr('A1'))).toEqual(0.5) //Should be 0.
```
[the remark this answers is not in the document]
Comment (izulin) [reply test/interpreter/function-npv.spec.ts:40]: For all of them. They treat cell reference as 1-length range (hence no-coerce), we treat reference as value (hence we coerce).

File: test/interpreter/function-pduration.spec.ts [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+      ['=PDURATION(2%, 12, 24)'],
+      ['=PDURATION(2%, -12, -24)'],
+      ['=PDURATION(0, 1, 1)'],
+    ])
+
+    expect(engine.getCellValue(adr('A1'))).toBeCloseTo(35.00278878, 6)
+    expect(engine.getCellValue(adr('A2'))).toEqual(detailedError(ErrorType.NUM, ErrorMessage.ValueSmall))
+    expect(engine.getCellValue(adr('A3'))).toEqual(detailedError(ErrorType.NUM, ErrorMessage.ValueSmall))
+  })
+})
```
[the remark this answers is not in the document]
Comment (izulin) [reply test/interpreter/function-pduration.spec.ts:26]: > Rate argument test missing for incorrect value. What if a negative value was passed? What if inncorect type was passed, ie. string? [2] "Constraints: Rate > 0; CurrentValue > 0; SpecifiedValue > 0"
> I don't see a test with #VALUE! error. But that condition is clearly described [1]: "If arguments are not using valid data types, PDURATION returns the #VALUE! error value."

That is because it is extremely simple function that fully uses our wrapper. Thus putting such test is testing the wrapper, not the implementation.

File: src/interpreter/plugin/FinancialPlugin.ts [hunk 9]
```diff
@@ -211,6 +212,45 @@ export class FinancialPlugin extends FunctionPlugin {
...
+      parameters: [
+        {argumentType: ArgumentTypes.NUMBER, greaterThan: 0},
+        {argumentType: ArgumentTypes.NUMBER, greaterThan: 0},
+        {argumentType: ArgumentTypes.NUMBER, greaterThan: 0},
+      ],
+    },
+    'XNPV': {
+      method: 'xnpv',
+      parameters: [
+        {argumentType: ArgumentTypes.NUMBER, greaterThan: 0},
```
Comment (bardek8) [inline src/interpreter/plugin/FinancialPlugin.ts:249]: not -1?

File: src/interpreter/plugin/FinancialPlugin.ts [hunk 10]
```diff
@@ -559,6 +599,118 @@ export class FinancialPlugin extends FunctionPlugin {
...
+        }
+        const dateArrNum = dateArr as number[]
+        if(dateArrNum.length !== valArrNum.length) {
+          return new CellError(ErrorType.NUM, ErrorMessage.EqualLength)
+        }
+        const n = dateArrNum.length
+        let ret = 0
+        for(let i=0;i<n;i++) {
+          dateArrNum[i] = Math.floor(dateArrNum[i])
+          if(dateArrNum[i]  < 0) {
```
Comment (bardek8) [inline src/interpreter/plugin/FinancialPlugin.ts:702]: Why do we need this? Isn't the subsequent check sufficient?

File: src/interpreter/plugin/FinancialPlugin.ts [hunk 11]
```diff
@@ -559,6 +599,118 @@ export class FinancialPlugin extends FunctionPlugin {
...
+    return this.runFunction(ast.args, formulaAddress, this.metadata('PDURATION'),
+      (rate: number, pv: number, fv: number) => (Math.log(fv) - Math.log(pv))/Math.log(1+rate)
+    )
+  }
+
+  public xnpv(ast: ProcedureAst, formulaAddress: SimpleCellAddress): InternalScalarValue {
+    return this.runFunction(ast.args, formulaAddress, this.metadata('XNPV'),
+      (rate: number, values: SimpleRangeValue, dates: SimpleRangeValue) => {
+        const valArr = values.valuesFromTopLeftCorner()
+        for(const val of valArr) {
```
Comment (bardek8) [inline src/interpreter/plugin/FinancialPlugin.ts:682]: don't we have to check that there is at least one positive and one negative value?

File: test/interpreter/function-mirr.spec.ts [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+      ['=MIRR(B3:E3,0.2,0.1)', -1, 0, -1, 1],
+      ['=MIRR(B3:E3,0.2,0.1)', -1, -1, 1],
+    ])
+    expect(engine.getCellValue(adr('A1'))).toEqual(3)
+    expect(engine.getCellValue(adr('A2'))).toEqual(0)
+    expect(engine.getCellValue(adr('A3'))).toBeCloseTo(-0.161201673643132, 6)
+    expect(engine.getCellValue(adr('A4'))).toBeCloseTo(-0.161201673643132, 6) // different value without 0
+  })
+  
+  it('should return #DIV/0! if at least one positive and one negative values condtion is not met', () => {
```
Comment (bardek8) [inline test/interpreter/function-mirr.spec.ts:29]:
```suggestion
  it('should return #DIV/0! if "contains at least one positive and one negative value" condition is not met', () => {
```

File: test/interpreter/function-mirr.spec.ts [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+    const engine = HyperFormula.buildFromArray([
+      ['=MIRR(B1:C1,1,1)', 1, -1],
+      ['=MIRR(B2:C2,1,1)', -1, 1],
+      ['=MIRR(B3:E3,0.2,0.1)', -1, 0, -1, 1],
+      ['=MIRR(B3:E3,0.2,0.1)', -1, -1, 1],
+    ])
+    expect(engine.getCellValue(adr('A1'))).toEqual(3)
+    expect(engine.getCellValue(adr('A2'))).toEqual(0)
+    expect(engine.getCellValue(adr('A3'))).toBeCloseTo(-0.161201673643132, 6)
+    expect(engine.getCellValue(adr('A4'))).toBeCloseTo(-0.161201673643132, 6) // different value without 0
```
Comment (bardek8) [inline test/interpreter/function-mirr.spec.ts:26]: Now A3=A4 - I believe this is not intended?
Then I would remove the comment, which might be slightly confusing.

File: src/interpreter/plugin/FinancialPlugin.ts [hunk 14]
```diff
@@ -211,6 +212,45 @@ export class FinancialPlugin extends FunctionPlugin {
...
+        {argumentType: ArgumentTypes.NUMBER},
+        {argumentType: ArgumentTypes.ANY},
+      ],
+      repeatLastArgs: 1
+    },
+    'MIRR': {
+      method: 'mirr',
+      parameters: [
+        {argumentType: ArgumentTypes.RANGE},
+        {argumentType: ArgumentTypes.NUMBER, greaterThan: -1},
```
Comment (bardek8) [inline src/interpreter/plugin/FinancialPlugin.ts:234]: Why do we require that? Actually everything is possible, only for -1 we return DIV0

File: test/interpreter/function-mirr.spec.ts [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+      ['=MIRR(1,1,1,1)'],
+    ])
+
+    expect(engine.getCellValue(adr('A1'))).toEqual(detailedError(ErrorType.NA, ErrorMessage.WrongArgNumber))
+    expect(engine.getCellValue(adr('A2'))).toEqual(detailedError(ErrorType.NA, ErrorMessage.WrongArgNumber))
+  })
+
+  it('should return correct value', () => {
+    const engine = HyperFormula.buildFromArray([
+      ['=MIRR(B1:C1,1,1)', 1, -1],
```
Comment (bardek8) [inline test/interpreter/function-mirr.spec.ts:18]: I would avoid using the same value for completely different parameters :) 
Can I also ask for the following test? 
`['=MIRR(B1:F1,0.3,-0.1)',1,2,-3,-5,8]`
(This will also check negative rates)

File: test/interpreter/function-npv.spec.ts [hunk 16]
```diff
@@ -0,0 +1,57 @@
...
+      ['=NPV(-1,1)'],
+    ])
+    expect(engine.getCellValue(adr('A1'))).toEqual(detailedError(ErrorType.NA))
+    expect(engine.getCellValue(adr('A2'))).toEqual(detailedError(ErrorType.VALUE, ErrorMessage.NumberCoercion))
+    expect(engine.getCellValue(adr('A3'))).toEqual(detailedError(ErrorType.NUM, ErrorMessage.NaN))
+  })
+
+
+  /**
+   * Discrepancy with products #1 and #2.
```
Comment (bardek8) [inline test/interpreter/function-npv.spec.ts:49]:
```suggestion
   * Inconsistency with products #1 and #2.
```

File: test/interpreter/function-npv.spec.ts [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+   */
+  it('cell reference', () => {
+    const engine = HyperFormula.buildFromArray([
+      ['=NPV(1,B1)', true]
+    ])
+    expect(engine.getCellValue(adr('A1'))).toEqual(0.5) //Should be 0.
```
Comment (bardek8) [inline test/interpreter/function-npv.spec.ts:55]:
```suggestion
    expect(engine.getCellValue(adr('A1'))).toEqual(0.5) //Both products #1 and #2 return 0 here.
```

File: test/interpreter/function-npv.spec.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (bardek8) [reply test/interpreter/function-npv.spec.ts:55]: Could you please also add this to known limitations? For example, sth like:
"we immediately instantiate references to single cells to their values instead of treating them as 1-length ranges, which slightly changes behavior of some functions (e.g. NPV)"

**Review by bardek8 — CHANGES_REQUESTED** [review]

File: src/interpreter/plugin/FinancialPlugin.ts (hunk 9 above)
Comment (izulin) [reply src/interpreter/plugin/FinancialPlugin.ts:249]: this is actual value implemented in XL

File: src/interpreter/plugin/FinancialPlugin.ts (hunk 14 above)
Comment (izulin) [reply src/interpreter/plugin/FinancialPlugin.ts:234]: ok, but then why does =MIRR({3,4,-3},-2,-2) results in NUM?

File: src/interpreter/plugin/FinancialPlugin.ts (hunk 11 above)
Comment (izulin) [reply src/interpreter/plugin/FinancialPlugin.ts:682]: in XL docs: yes. in XL implementation: no

File: src/interpreter/plugin/FinancialPlugin.ts [hunk 18]
```diff
@@ -591,3 +743,15 @@ function fvCore(rate: number, periods: number, payment: number, value: number, t
 function ppmtCore(rate: number, period: number, periods: number, present: number, future: number, type: number): number {
   return pmtCore(rate, periods, present, future, type) - ipmtCore(rate, period, periods, present, future, type)
 }
+
+function npvCore(rate: number, args: number[]): number {
+  let acc = 0
+  for(let i=args.length-1;i>=0;i--) {
+    acc += args[i]
+    if(acc===0 && rate===-1) {
```
[the remark this answers is not in the document]
Comment (izulin) [reply src/interpreter/plugin/FinancialPlugin.ts:751]: We have a test for that. If acc is not 0 and rate == -1, then here it would evaluate implicitly to NUM from NaN.
I can change it to explicit DIV0.

File: test/interpreter/function-fvschedule.spec.ts (hunk 5 above)
Comment (izulin) [reply test/interpreter/function-fvschedule.spec.ts:36]: Can be negative. It is not a border case.

File: test/interpreter/function-mirr.spec.ts (hunk 13 above)
Comment (izulin) [reply test/interpreter/function-mirr.spec.ts:26]: right, copy-paste gone wrong

**Review by bardek8 — APPROVED** [review]
Thanks for the fixes! It looks good to me, I'm passing the token to  @wojciechczerniak .

File: docs/guide/known-limitations.md [hunk 19]
```diff
@@ -34,4 +34,5 @@ you can't compare the arguments in a formula like this:
   * Asynchronous functions
   * Structured references ("Tables")
   * Currency data type
-  * Relative named expressions
\ No newline at end of file
+  * Relative named expressions
+* We immediately instantiate references to single cells to their values instead of treating them as 1-length ranges, which slightly changes behavior of some functions (e.g. NPV).
```
Comment (wojciechczerniak) [inline docs/guide/known-limitations.md:38]:
```suggestion
  * Relative named expressions
  * We immediately instantiate references to single cells to their values instead of treating them as 1-length ranges, which slightly changes behavior of some functions (e.g. NPV).
```

File: test/interpreter/function-fvschedule.spec.ts (hunk 5 above)
Comment (wojciechczerniak) [reply test/interpreter/function-fvschedule.spec.ts:36]: Ok, thx

File: test/interpreter/function-npv.spec.ts [hunk 20]
```diff
@@ -0,0 +1,57 @@
...
+    ])
+
+    expect(engine.getCellValue(adr('A1'))).toEqualError(detailedError(ErrorType.NA, ErrorMessage.WrongArgNumber))
+  })
+
+  it('should return correct value', () => {
+    const engine = HyperFormula.buildFromArray([
+      ['=NPV(1, B1:C1)', 1, 'abcd'],
+      ['=NPV(1, B2:C2)', true, 1],
+      ['=NPV(1, TRUE(), 1)'],
```
Comment (wojciechczerniak) [inline test/interpreter/function-npv.spec.ts:18]: Excel documentation says:

> Arguments that are empty cells, logical values, or text representations of numbers, error values, or text that cannot be translated into numbers are ignored.
>
> If an argument is an array or reference, only numbers in that array or reference are counted. Empty cells, logical values, text, or error values in the array or reference are ignored.

But values returned are as in Excel on Mac. Google Spreadsheet has a different opinion:
Screenshot 2020-10-14 at 19 30 47

Comment (wojciechczerniak) [reply test/interpreter/function-npv.spec.ts:18]: I would split into two scenarios then
```
it('should ignore logical and text values', () => {
      ['=NPV(1, B1:C1)', 1, 'abcd'],
      ['=NPV(1, B2:C2)', true, 1],
```

```
it('should be compatible with Excel ?', () => {
      ['=NPV(1, TRUE(), 1)'],
```

Should we add HF.NPV that will calculate this correctly or move this implementation to MS.NPV and implement this correctly as ODFF and GS does?

File: test/interpreter/function-pduration.spec.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (wojciechczerniak) [reply test/interpreter/function-pduration.spec.ts:34]: 👍

File: test/interpreter/function-npv.spec.ts (hunk 3 above)
Comment (wojciechczerniak) [reply test/interpreter/function-npv.spec.ts:35]: Thx 👍

**Review by wojciechczerniak — CHANGES_REQUESTED** [review]
Awesome 💪 There is one problem with NPV function though. Maybe you have more context, WDYT?

File: test/interpreter/function-npv.spec.ts (hunk 20 above)
Comment (izulin) [reply test/interpreter/function-npv.spec.ts:18]: I think we might mark the inconsistency between implementations. Marking every function where we found it is ok. Implementing two versions of each might be too much (for now).

File: docs/guide/known-limitations.md [hunk 21]
```diff
@@ -35,5 +35,9 @@ you can't compare the arguments in a formula like this:
   * Structured references ("Tables")
   * Currency data type
   * Relative named expressions
+<<<<<<< HEAD
+  * We immediately instantiate references to single cells to their values instead of treating them as 1-length ranges, which slightly changes behavior of some functions (e.g. NPV).
+=======
   * Functions cannot use UI metadata (e.g. hidden rows for SUBTOTAL).
   * SUBTOTAL function does not ignore nested subtotals.
+>>>>>>> develop
```
Comment (wojciechczerniak) [inline docs/guide/known-limitations.md:43]: Merge problem

**Review by wojciechczerniak — CHANGES_REQUESTED** [review]
All ok, but we can't merge this until git comments are removed

**Review by wojciechczerniak — APPROVED** [review]
LGreatTM 🏆
